File: src/limitedwip/common/settings/LimitedWIPSettings.java
package limitedwip.common.settings;

import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.util.Range;
import com.intellij.util.xmlb.XmlSerializerUtil;
import limitedwip.common.PluginId;

@State(
	name = PluginId.value + "Settings",
	storages = @Storage(file = "$APP_CONFIG$/limitedwip.ui.settings.xml")
)
public class LimitedWIPSettings implements PersistentStateComponent<LimitedWIPSettings>  {
	public static final Range<Integer> minutesToRevertRange = new Range<Integer>(1, 99);
	public static final Range<Integer> changedLinesRange = new Range<Integer>(1, 999);
	public static final Range<Integer> notificationIntervalRange = new Range<Integer>(1, 99);

	public boolean autoRevertEnabled = false;
	public int minutesTillRevert = 2;
	public boolean notifyOnRevert = true;
	public boolean showTimerInToolbar = true;

	public boolean watchdogEnabled = true;
	public int maxLinesInChange = 80;
	public int notificationIntervalInMinutes = 1;
	public boolean disableCommitsAboveThreshold = false;
	public boolean showRemainingChangesInToolbar = true;


	public int secondsTillRevert() {
		return minutesTillRevert * 60;
	}

	public int notificationIntervalInSeconds() {
		return notificationIntervalInMinutes * 60;
	}

	@Override public LimitedWIPSettings getState() {
		return this;
	}

	@Override public void loadState(LimitedWIPSettings state) {
		XmlSerializerUtil.copyBean(state, this);
	}

	@SuppressWarnings("RedundantIfStatement") @Override public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		LimitedWIPSettings settings = (LimitedWIPSettings) o;

		if (autoRevertEnabled != settings.autoRevertEnabled) return false;
		if (disableCommitsAboveThreshold != settings.disableCommitsAboveThreshold) return false;
		if (maxLinesInChange != settings.maxLinesInChange) return false;
		if (minutesTillRevert != settings.minutesTillRevert) return false;
		if (notificationIntervalInMinutes != settings.notificationIntervalInMinutes) return false;
		if (notifyOnRevert != settings.notifyOnRevert) return false;
		if (showRemainingChangesInToolbar != settings.showRemainingChangesInToolbar) return false;
		if (showTimerInToolbar != settings.showTimerInToolbar) return false;
		if (watchdogEnabled != settings.watchdogEnabled) return false;

		return true;
	}

	@Override public int hashCode() {
		int result = (autoRevertEnabled ? 1 : 0);
		result = 31 * result + minutesTillRevert;
		result = 31 * result + (showTimerInToolbar ? 1 : 0);
		result = 31 * result + (watchdogEnabled ? 1 : 0);
		result = 31 * result + maxLinesInChange;
		result = 31 * result + notificationIntervalInMinutes;
		result = 31 * result + (disableCommitsAboveThreshold ? 1 : 0);
		result = 31 * result + (notifyOnRevert ? 1 : 0);
		result = 31 * result + (showRemainingChangesInToolbar ? 1 : 0);
		return result;
	}

	@Override public String toString() {
		return "Settings{" +
				"autoRevertEnabled=" + autoRevertEnabled +
				", minutesTillRevert=" + minutesTillRevert +
				", showTimerInToolbar=" + showTimerInToolbar +
				", watchdogEnabled=" + watchdogEnabled +
				", maxLinesInChange=" + maxLinesInChange +
				", notificationIntervalInMinutes=" + notificationIntervalInMinutes +
				", disableCommitsAboveThreshold=" + disableCommitsAboveThreshold +
				", notifyOnRevert=" + notifyOnRevert +
				", showRemainingChangesInToolbar=" + showRemainingChangesInToolbar +
				'}';
	}

}

File: src/limitedwip/common/settings/LimitedWipSettings.kt
package limitedwip.common.settings

import com.intellij.openapi.Disposable
import com.intellij.openapi.components.PersistentStateComponent
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Disposer
import com.intellij.util.xmlb.XmlSerializerUtil
import limitedwip.common.settings.CommitMessageSource.LastCommit
import limitedwip.common.settings.LimitedWipSettings.Companion.never
import limitedwip.common.settings.TcrAction.Commit
import limitedwip.common.settings.TimeUnit.Minutes

@State(name = "LimitedWIPPluginSettings", storages = [Storage(value = "limited-wip.xml")])
data class LimitedWipSettings(
    var watchdogEnabled: Boolean = true,
    var maxLinesInChange: Int = 80,
    var notificationIntervalInMinutes: Int = never,
    var noCommitsAboveThreshold: Boolean = false,
    var showRemainingChangesInToolbar: Boolean = true,
    var exclusions: String = "",

    var autoRevertEnabled: Boolean = false,
    var timeTillRevert: Int = 2,
    var timeUnitTillRevert: TimeUnit = Minutes,
    var notifyOnRevert: Boolean = true,
    var showTimerInToolbar: Boolean = true,

    var tcrEnabled: Boolean = false,
    var tcrActionOnPassedTest: TcrAction = Commit,
    var commitMessageSource: CommitMessageSource = LastCommit,
    var notifyOnTcrRevert: Boolean = false,
    var doNotRevertTests: Boolean = false,
    var doNotRevertFiles: String = ""
): PersistentStateComponent<LimitedWipSettings> {
    private val listeners = ArrayList<Listener>()

    override fun getState() = this

    override fun loadState(state: LimitedWipSettings) {
        XmlSerializerUtil.copyBean(state, this)
        notifyListeners()
    }

    fun addListener(parentDisposable: Disposable, listener: Listener) {
        listeners.add(listener)
        Disposer.register(parentDisposable) { listeners.remove(listener) }
    }

    fun notifyListeners() {
        listeners.forEach { it.onUpdate(this) }
    }

    interface Listener {
        fun onUpdate(settings: LimitedWipSettings)
    }

    companion object {
        const val never = Int.MAX_VALUE
        private val timeTillRevertRange = IntRange(1, 999)
        private val changedLinesRange = IntRange(1, 999)
        private val notificationIntervalRange = IntRange(1, never)

        fun isValidTimeTillRevert(duration: Int) = duration in timeTillRevertRange
        fun isValidChangedSizeRange(lineCount: Int) = lineCount in changedLinesRange
        fun isValidNotificationInterval(interval: Int) = interval in notificationIntervalRange

        fun getInstance(project: Project): LimitedWipSettings =
            project.getService(LimitedWipSettings::class.java)
    }
}

enum class TcrAction {
    OpenCommitDialog,
    Commit,
    AmendCommit,
    CommitAndPush
}

enum class CommitMessageSource {
    LastCommit,
    ChangeListName
}

enum class TimeUnit {
    Seconds,
    Minutes;

    fun toSeconds(duration: Int): Int =
        if (duration == never) never
        else when (this) {
            Seconds -> duration
            Minutes -> duration * 60
        }
}